tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,518 @@
+
+  For each graph, the partitioning algorithm takes in the graph's `GraphDef`
+  proto and output names, partitions the graph, and returns a list of
+  ExecutionSpecs. Later, the beam_pipeline library can take in the
+  ExecutionSpecs and execute the partitioned subgraphs.
+
+  Args:
+    graph_name_to_graph_def: A mapping from graph names to `GraphDef` protos.
+    graph_name_to_output_names: A mapping from graph names to lists of their
+                                output node names.
+
+  Returns:
+    A mapping from graph names to a list of ExecutionSpecs, where the order
+    of the list represents the order of execution.
+  """
+  graph_name_to_specs = {}
+  for graph_name in graph_name_to_graph_def:
+    specs = _partition_one_graph(graph_name_to_graph_def[graph_name],
+                                 graph_name_to_output_names[graph_name])
+    graph_name_to_specs[graph_name] = specs
+  return graph_name_to_specs
+
+
+def _partition_one_graph(
+    graph_def: graph_pb2.GraphDef,
+    output_names: List[Text]) -> List[ExecutionSpec]:
+  """Partitions one graph.
+
+  Args:
+    graph_def: A `GraphDef` proto for that graph.
+    output_names: A list of graph's output node names.
+
+  Returns:
+    A list of ExecutionSpecs.
+  """
+  graph = _get_graph(graph_def)
+  node_name_to_node_def = _get_node_name_to_node_def(graph_def)
+
+  remote_op_to_parents = _get_remote_op_to_parents(node_name_to_node_def)
+
+  specs = _get_execution_specs(graph_def,
+                               output_names,
+                               graph,
+                               node_name_to_node_def,
+                               remote_op_to_parents)
+
+  _modify_execution_specs(specs)
+
+  return specs
+
+
+def _get_graph(graph_def: graph_pb2.GraphDef) -> tf.Graph:
+  with tf.compat.v1.Session(graph=tf.Graph()) as sess:
+    tf.import_graph_def(graph_def)
+    return sess.graph
+
+
+def _get_node_name_to_node_def(
+    graph_def: graph_pb2.GraphDef) -> Dict[Text, node_def_pb2.NodeDef]:
+  return {node.name: node for node in graph_def.node}
+
+
+def _get_remote_op_to_parents(
+    node_name_to_node_def: Mapping[Text, node_def_pb2.NodeDef]
+) -> Dict[Text, List[Text]]:
+  """Gets the execution dependencies between remote ops.
+
+  The remote op parents must be executed before executing a remote op.
+
+  Args:
+    node_name_to_node_def: A mapping from node names to `NodeDef` protos.
+
+  Returns:
+    A mapping from a remote op name to a list of immediate remote op parent
+    names.
+  """
+  remote_op_to_parents = {}
+
+  for node in node_name_to_node_def.values():
+    if _is_remote_op(node):
+      remote_op_to_parents[node.name] = _get_remote_op_parents(
+          node.name, node_name_to_node_def)
+
+  return remote_op_to_parents
+
+
+def _get_remote_op_parents(
+    remote_op_name: Text,
+    node_name_to_node_def: Mapping[Text, node_def_pb2.NodeDef]
+) -> List[Text]:
+  """Finds the immediate remote op parents for a remote op.
```

Comment:
Could you use the term "immediate dependencies" rather than "parents"? The term "parents" is ambiguous.

Consider the following:
```
a*
|
b   c*
 \  /
  d*
```
where `*` indicates remote ops. Here the immediate remote op dependencies of `d*` are `a*` and `c*`. However the parents of `d*` in the graph are `b` and `c*`.

Reply:
Addressed.